python/__init__.py

```diff
@@ -72,6 +72,27 @@ def __hash__(self):
 except ImportError:
     pass
 
+# Define a `with edit(layer)` statement
+
+class edit:
+    def __init__(self,layer):
+        self.layer = layer
+
+    def __enter__(self):
+        assert self.layer.startEditing()
+        return self.layer
+
+    def __exit__(self, ex_type, ex_value, traceback):
+        if ex_type is None:
+            assert self.layer.commitChanges()
```

Comment:
Oh I didn't mean define it within the edit class. Just use it there. Make a
new exception class. Or am I missing what you mean.

On Wed, Aug 12, 2015 at 10:02 PM, Matthias Kuhn <notifications@github.com>
wrote:

> In python/__init__.py
> <https://github.com/qgis/QGIS/pull/2242#discussion_r36851407>:
>
> > @@ -72,6 +72,27 @@ def __hash__(self):
> >  except ImportError:
> >      pass
> >
> > +# Define a `with edit(layer)` statement
> > +
> > +class edit:
> > +    def __init__(self,layer):
> > +        self.layer = layer
> > +
> > +    def __enter__(self):
> > +        assert self.layer.startEditing()
> > +        return self.layer
> > +
> > +    def __exit__(self, ex_type, ex_value, traceback):
> > +        if ex_type is None:
> > +            assert self.layer.commitChanges()
>
> If it's defined in sip we will get the exception even if not using with
> (like a lot of code currently does). I think that's a big plus.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/qgis/QGIS/pull/2242/files#r36851407>.
>